src/generate_csv.py

```diff
@@ -0,0 +1,86 @@
+import boto3
+from .utils import ensure, first, splitfilter, select_keys, keys
+import time
+import sys
+import os
+import csv
+
+def client():
+    return boto3.client('iam')
+
+def coerce(row):
+    value_lookups = {
+        'false': False,
+        'true': True,
+        'N/A': None,
+        'no_information': None # wtf? treat as N/A
+    }
+    return {key: value_lookups.get(val, val) for key, val in row.items()}
+
+def load_csv(path):
+    with open(path, 'r') as csvfile:
+        reader = csv.DictReader(csvfile)
+        return [coerce(row) for row in reader]
+
+def dump_csv(path, list_of_dicts):
+    with open(path, 'w') as csvfile:
+        fieldnames = keys(first(list_of_dicts))
+        writer = csv.DictWriter(csvfile, fieldnames)
+        writer.writeheader()
+        for row in list_of_dicts:
+            writer.writerow(row)
+    return path
+
+def num_uppercase(string):
+    return sum(1 for char in string if char.isupper())
+
+# ---
+
+def generate_credential_report():
+    iam = client()
+    print('requesting report')
+    while True:
+        sys.stdout.write('polling ... ')
+        resp = iam.generate_credential_report()
+        if resp not in ['STARTED', 'INPROGRESS']:
+            print('done.')
+            break
+        time.sleep(2) # seconds
```

Comment:
a time limit to avoid infinite looping?

Reply:
> if resp not in ['STARTED', 'INPROGRESS']:

This line is wrong, it's never doing a second loop. It was working for me because I called generate in the repl and it was always there subsequently. I'll fix it

The value it was checking can only ever be one of three values. If it's not 'started' or 'inprogress', it breaks the loop. No other guard necessary

Reply:
fixed